crates/test/src/rt/mod.rs

```diff
@@ -0,0 +1,379 @@
+//! Internal-only runtime module used for the `wasm_bindgen_test` crate.
+//!
+//! No API contained in this module will respect semver, these should all be
+//! considered private APIs.
+
+// # Architecture of `wasm_bindgen_test`
+//
+// This module can seem a bit funky, but it's intended to be the runtime support
+// of the `#[wasm_bindgen_test]` macro and be amenable to executing wasm test
+// suites. The general idea is that for a wasm test binary there will be a set
+// of functions tagged `#[wasm_bindgen_test]`. It's the job of the runtime
+// support to execute all of these functions, collecting and collating the
+// results.
+//
+// This runtime support works in tandem with the `wasm-bindgen-test-runner`
+// binary as part of the `wasm-bindgen-cli` package.
+//
+// ## High Level Overview
+//
+// Here's a rough and (semi) high level overview of what happens when this crate
+// runs.
+//
+// * First, the user runs `cargo test --target wasm32-unknown-unknown`
+//
+// * Cargo then compiles all the test suites (aka `tests/*.rs`) as wasm binaries
+//   (the `bin` crate type). These binaries all have entry points that are
+//   `main` functions, but it's actually not used. The binaries are also
+//   compiled with `--test`, which means they're linked to the standard `test`
+//   crate, but this crate doesn't work on wasm and so we bypass it entirely.
```

Comment:
Ah yeah those two properties though were why I consider it as "not working" in the sense that failing on the first failure isn't great for a test harness and similarly not being able to output anything is a bummer :(

Reply:
It is though nicer to have regular tests in case of developing a library that should work on both wasm and natively, even if limited like this.

I remember there is like wasm_syscall feature in std that should be able to enable printing, not sure about unwind.

Reply:
It would be nice to use `#[test]`, yeah, but customizing that and getting it working on stable is a long way off. The `wasm_syscall` feature is unlikely to ever be stabilized